File: modules/yandex-map/ios/YandexMapModule.swift
import ExpoModulesCore

public class YandexMapModule: Module {
  public func definition() -> ModuleDefinition {
    Name("YandexMap")

    View(YandexMapView.self) {
        Events("onMapReady", "onMapTap", "onMapLongTap", "onClusterPress")
        
        Prop("clusterized") { (view: YandexMapView, isClusterized: Bool) in
            view.setClusterized(isClusterized)
        }
        
        Prop("clusterStyle") { (view: YandexMapView, clusterStyle: ClusterStyleModel) in
            view.setClusterStyle(clusterStyle)
        }
        
        Prop("clusterConfig") { (view: YandexMapView, config: ClusterConfigModel) in
            view.setClusterConfig(config)
        }
        
        Prop("scrollEnabled") { (view: YandexMapView, isScrollEnabled: Bool) in
            view.setScrollEnabled(isScrollEnabled)
        }
        
        Prop("zoomEnabled") { (view: YandexMapView, isZoomEnabled: Bool) in
            view.setZoomEnabled(isZoomEnabled)
        }
        
        Prop("rotationEnabled") { (view: YandexMapView, isRotationEnabled: Bool) in
            view.setRotationEnabled(isRotationEnabled)
        }
        
        Prop("tiltEnabled") { (view: YandexMapView, isTiltEnabled: Bool) in
            view.setTiltEnabled(isTiltEnabled)
        }
        
        Prop("initialRegion") { (view: YandexMapView, position: Position) in
            view.setInitialRegion(position)
        }
        
        AsyncFunction("moveToCenter") { (view: YandexMapView, position: Position) in
            view.moveCamera(position)
        }
        
        AsyncFunction("fitMarkers") { (view: YandexMapView, points: Array<Position>) in
            let markers = points.map { $0.toPoint() }
            
            view.fitMarkers(markers)
        }
    }
  }
}

File: modules/yandex-map/android/src/main/java/expo/modules/yandexmap/modules/YandexMapModule.kt
package expo.modules.yandexmap.modules

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.yandexmap.model.ClusterConfigModel
import expo.modules.yandexmap.model.ClusterStyleModel
import expo.modules.yandexmap.model.FocusRect
import expo.modules.yandexmap.model.Position
import expo.modules.yandexmap.view.YandexMapView

class YandexMapModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("YandexMap")

        View(YandexMapView::class) {
            Events("onMapReady", "onClusterPress", "onMapTap", "onMapLongTap")

            Prop("clusterized") { view: YandexMapView, clusterized: Boolean ->
                view.setClusterizedValue(clusterized)
            }

            Prop("clusterStyle") { view: YandexMapView, style: ClusterStyleModel ->
                view.setClusterStyleValue(style.toClusterStyle())
            }

            Prop("clusterConfig") { view: YandexMapView, config: ClusterConfigModel ->
                view.setClusterConfigValue(config)
            }

            Prop("initialRegion") { view: YandexMapView, position: Position ->
                view.setInitialRegion(position)
            }

            Prop("scrollEnabled") { view: YandexMapView, isEnabled: Boolean ->
                view.setScrollEnabled(isEnabled)
            }

            Prop("zoomEnabled") { view: YandexMapView, isEnabled: Boolean ->
                view.setZoomEnabled(isEnabled)
            }

            Prop("rotationEnabled") { view: YandexMapView, isEnabled: Boolean ->
                view.setRotationEnabled(isEnabled)
            }

            Prop("tiltEnabled") { view: YandexMapView, isEnabled: Boolean ->
                view.setTiltEnabled(isEnabled)
            }
            // TODO: implement after focus on object
            Prop("focusRect") { view: YandexMapView, rect: FocusRect ->
                val tlX = rect.topLeft.x
                val tlY = rect.topLeft.y
                val brX = rect.bottomRight.x
                val brY = rect.bottomRight.y

                view.setFocusRect(tlX, tlY, brX, brY)
            }

            AsyncFunction("moveToCenter") { view: YandexMapView, position: Position ->
                view.moveCamera(position)
            }

            AsyncFunction("fitMarkers") { view: YandexMapView, points: List<Position> ->
                val markers = points.map { it.toPoint() }

                view.fitMarkers(markers)
            }
        }
    }
}
